crates/sui-indexer/src/indexer_reader.rs

```diff
@@ -238,6 +253,188 @@ impl IndexerReader {
+                    boxed_query =
+                        boxed_query.filter(checkpoints::sequence_number.lt(cursor as i64));
+                } else {
+                    boxed_query =
+                        boxed_query.filter(checkpoints::sequence_number.gt(cursor as i64));
+                }
+            }
+            if descending_order {
+                boxed_query = boxed_query.order_by(checkpoints::sequence_number.desc());
+            } else {
+                boxed_query = boxed_query.order_by(checkpoints::sequence_number.asc());
+            }
+
+            boxed_query
+                .limit(limit as i64)
+                .load::<StoredCheckpoint>(conn)
+        })
+    }
+
+    pub fn get_checkpoints(
+        &self,
+        cursor: Option<u64>,
+        limit: usize,
+        descending_order: bool,
+    ) -> Result<Vec<sui_json_rpc_types::Checkpoint>, IndexerError> {
+        self.get_checkpoints_from_db(cursor, limit, descending_order)?
+            .into_iter()
+            .map(sui_json_rpc_types::Checkpoint::try_from)
+            .collect()
+    }
+
+    pub fn get_transaction(
+        &self,
+        digest: TransactionDigest,
+    ) -> Result<Option<StoredTransaction>, IndexerError> {
+        self.run_query(|conn| {
+            transactions::table
+                .filter(transactions::transaction_digest.eq(digest.into_inner().to_vec()))
+                .first::<StoredTransaction>(conn)
+                .optional()
+        })
+    }
+
+    pub fn multi_get_transactions(
+        &self,
+        digests: &[TransactionDigest],
+    ) -> Result<Vec<StoredTransaction>, IndexerError> {
+        let digests = digests
+            .iter()
+            .map(|digest| digest.inner().to_vec())
+            .collect::<Vec<_>>();
+        self.run_query(|conn| {
+            transactions::table
+                .filter(transactions::transaction_digest.eq_any(digests))
+                .load::<StoredTransaction>(conn)
+        })
+    }
+
+    pub fn multi_get_transaction_block_response(
+        &self,
+        digests: &[TransactionDigest],
+        options: &sui_json_rpc_types::SuiTransactionBlockResponseOptions,
+    ) -> Result<Vec<sui_json_rpc_types::SuiTransactionBlockResponse>, IndexerError> {
+        self.multi_get_transactions(digests)?
+            .into_iter()
+            .map(|transaction| transaction.try_into_sui_transaction_block_response(options, self))
+            .collect::<Result<Vec<_>, _>>()
+    }
+
+    pub async fn multi_get_transaction_block_response_async(
+        &self,
+        digests: Vec<TransactionDigest>,
+        options: sui_json_rpc_types::SuiTransactionBlockResponseOptions,
+    ) -> Result<Vec<sui_json_rpc_types::SuiTransactionBlockResponse>, IndexerError> {
+        self.spawn_blocking(move |this| {
+            this.multi_get_transaction_block_response(&digests, &options)
+        })
+        .await
+    }
+
+    //TODO can we only use the events table?
```

Comment:
if `events` is sill a column in `transactions`, it's more efficient to read it from `transactions` table.
technically, `events` column and `events` table store roughly the same data, it's debatable if we want to remove `events` column. Other than here, it's also more efficient reading `SuiTransactionBlockResponse` with the extra `events` column

Reply:
The main motivation of having events column in transactions table is that once we offload cold data to remote storage, everything is still in one place. Otherwise to access them we may have to query multiple data sources which is not great.